### ISSUE TITLE
Tools comparisson (dts-bundle-generator, api-extractor, rollup-plugin-dts, npm-dts)

### ISSUE THREAD

**Issue by octogonz** [timocov/dts-bundle-generator#68]
@timocov FYI we've been working on a tool [API Extractor](https://api-extractor.com/pages/basics/intro/) that seems to have similar goals to **dts-bundle-generator**.  It might be interesting to compare the features and goals of the two projects, and see if anything can be shared between them.

**Comment by timocov** [comment]
Hi @pgonzal,

thank you for your interest in the project!

After slight looking up I see the next differences:

1. `dts-bundle-generator` does not require configure anything - you can just run it via command line and it'll just work. But you're able to configure some stuff if you need it (see https://github.com/timocov/dts-bundle-generator#usage).
1. To run `dts-bundle-generator` you don't need to compile your project - it'll do it by itself while generating (and it does it in the memory, so no unnecessary output will be written to your drive).
1. It looks like `api-extractor` requires you export all interfaces/functions/const/etc from your entry point if you want to export them, and doesn't follow dependencies, right? If so, this is difference also, because `dts-bundle-generator` will follow dependencies and put them all into result file.
1. `api-extractor` requires add tag to an export member (`@public`, `@internal`, `@beta`).
1. Installing `api-extractor` adds `@types/node` to the project and it may affect the build (in my case it did it). I'm not sure that it's a good idea to put it to `dependencies` instead of `devDependencies`.
1. `api-extractor` requires TypeScript version `3.1.x`, but `dts-bundle-generator` works with any versions of TypeScript starts with `2.6.1`.
1. It looks like configuring `api-extractor` might be painful. I still have an error (at least it looks red) `The "mainDtsRollupPath" setting was not specified. In this case, the package.json "typings" field must point to a file under the "publishFolder": ./dist`, but I've specified `typings` field in the `package.json` file and don't know what it actually wants from me.
1. With `dts-bundle-generator` you can generate dts bundle for several entry points ([example](https://github.com/timocov/dts-bundle-generator/blob/d8c7f42daa3c0856a1c4f573d6a3b6ac5f43fc43/dts-config.js#L8-L17)), it looks like `api-extractor` supports only one entry point per project.

Correct me if I were wrong somewhere about `api-extractor`. Also, it would be nice to see what do you think about it and what differences (or pros and cons) you see between projects.

**Comment by pgonzal** [comment]
Thanks for writing up this comparison!  It's very interesting.

> 2\. To run `dts-bundle-generator` you don't need to compile your project - it'll do it by itself while generating (and it does it in the memory, so no unnecessary output will be written to your drive).

Right, API Extractor processes .d.ts files instead of .ts files.  This supports the usage of preprocessors (e.g. a .js -> .d.ts generator), and also simplifies the processing that we do, since .d.ts files are in a normalized form

> 3\. It looks like `api-extractor` requires you export all interfaces/functions/const/etc from your entry point if you want to export them, and doesn't follow dependencies, right? If so, this is difference also, because `dts-bundle-generator` will follow dependencies and put them all into result file.

What do you mean by "follow dependencies"?  For locally referenced files `import { X } from './X';` it does follow and collapse these.  For external NPM packages `import { X } from 'package';` it emits the import statement in the rollup .d.ts.  Inlining such a declaration seems like it would be incorrect behavior.

> 4\. `api-extractor` requires add tag to an export member (`@public`, `@internal`, `@beta`).

This can be configured to be optional.

> 5\. Installing `api-extractor` adds `@types/node` to the project and it may affect the build (in my case it did it). I'm not sure that it's a good idea to put it to `dependencies` instead of `devDependencies`.

Wow, thanks for pointing this out!  We overlooked it for a long time.  :-)

> 6\. `api-extractor` requires TypeScript version `3.1.x`, but `dts-bundle-generator` works with any versions of TypeScript starts with `2.6.1`.

API Extractor installs with a specific (very recent) TypeScript package that it uses for its own analysis, but this is typically different from the version that your project builds with.  For example when processing a **2.6.1** project, it can use the **2.6.1** runtime library typings while still invoking the 3.1 engine.

> 7\. It looks like configuring `api-extractor` might be painful. I still have an error (at least it looks red) `The "mainDtsRollupPath" setting was not specified. In this case, the package.json "typings" field must point to a file under the "publishFolder": ./dist`, but I've specified `typings` field in the `package.json` file and don't know what it actually wants from me.

Yeah the configuration is a weak area, since in our own usage at Microsoft, all the projects are in a huge monorepo with a standardized configuration, so nobody customizes the configuration directly. The external community has complained about this, and it's a high priority to improve the configuration and make it more flexible.

> 8\. With `dts-bundle-generator` you can generate dts bundle for several entry points ([example](https://github.com/timocov/dts-bundle-generator/blob/d8c7f42daa3c0856a1c4f573d6a3b6ac5f43fc43/dts-config.js#L8-L17)), it looks like `api-extractor` supports only one entry point per project.

That's true. The analyzer engine partially supports path-based entry points, but they are very difficult to support in general. (How does the documentation system represent them? When generating a .d.ts rollup, where do you put common local declarations that are shared by two different entry points?)

**Comment by timocov** [comment]
> Right, API Extractor processes .d.ts files instead of .ts files.

AFAIK some `dts-bundle-generator`'s users use it with their `d.ts` files and it works fine. Actually there is no restrictions to use it only with `.ts` files, but it is just simpler in a common case.

>  For locally referenced files import { X } from './X'; it does follow and collapse these. 

Oh, great! It is quite possible that I just didn't get it work in a proper way 🙁

> This requirement can be disabled in the config file.

👌 Will know, thanks!

>  For example when processing a 2.6.1 project, it can use the 2.6.1 runtime library typings while still invoking the 3.1 engine.

In my case, I've tried to run in a project with 3.2.1 TypeScript, and it tried to use `lib` files from `3.1.6`, but they're incompatible and I get a lot or errors about it. As temporarily fix I've enabled `skipLibCheck` compiler option, but this isn't what I want to have in the compiler options 🙂

> The external community has complained about this

Another, but the similar opinion about configuration - https://github.com/Swatinem/rollup-plugin-dts#why.

> How should a documentation web site represent them?

What do you mean "documentation web site"? It is about the documentation site, which your tool generates? If so, I don't know the answer 🙁 In my case I just pointing to separate (not main) file in docs (see [JS file example](https://github.com/timocov/dts-bundle-generator/tree/master/src/config-file#tips)).

> When generating a .d.ts rollup, where do you emit common local declarations that are shared by two different entry points?

I haven't think about it yet deeply, but it is possible, that it is not a big problem in a common case. In a project where I'm working on we generate 3 different d.ts bundles (and there is a lot of code in a common between them), and it seems that everything is ok for now with that.

**Comment by timocov** [comment]
> where do you emit common local declarations that are shared by two different entry points?

It sounds like a good new issue 🙂 #69

**Comment by Swatinem** [comment]
Hi, and let me also chime in as the maintainer of [rollup-plugin-dts](https://github.com/Swatinem/rollup-plugin-dts)

As I have also mentioned in my README, I have tried `api-extractor` as well, but my biggest hurdle was also the configuration as @timocov also mentioned. In the end I couldn’t get it to work at all and just gave up.

So I have nothing to actually compare the *output quality and differences*, like @timocov asked me to here: https://github.com/Swatinem/rollup-plugin-dts/issues/5

But apart from the configuration issue that was already mentioned, my biggest beef is that it requires me to create intermediate output in the form of pre-rollup `.d.ts` files (point **2.** of @timocov’s list), which goes against how I would like my library workflows to work. (use `tsc` for typechecking only, and use a bundler like rollup / webpack to actually generate the output; I don’t want to have to manage/clean any intermediate build artifacts)

Also concerning **4.**, I recently found out that `tsc` has a `compilerOptions` for that which I plan to rely on at some point (https://github.com/Swatinem/rollup-plugin-dts/issues/2)

**Comment by octogonz** [comment]
`compilerOptions.stripInternal` is very limited.  Last time I tried it, if you did this...

```ts
/**
 * For more info, please email bob@internalideas.com
 */
export class Base {
}
export class Child extends Base {
}
```

...it would strip `Base` because it found the `@internal` substring in there, resulting in a `Child` definition that does not compile.  Whereas API Extractor relies on the [TSDoc](https://github.com/Microsft/tsdoc) grammar to understand that `@internal` is not a tag in this context.  Also, if something is marked as `@internal` that is referenced by an `@public` or `@beta` declaration, API Extractor will detect this and issue a warning about inconsistent visibility.

> As I have also mentioned in my README, I have tried `api-extractor` as well, but my biggest hurdle was also the configuration as @timocov also mentioned.

This is useful feedback. Thanks!

> But apart from the configuration issue that was already mentioned, my biggest beef is that it requires me to create intermediate output in the form of pre-rollup `.d.ts` files

It would be possible to pass these as in-memory data structures rather than writing them to disk. Someone else proposed this in https://github.com/Microsoft/web-build-tools/issues/1010. (It's not a very high priority for me personally, since from an opposite perspective I find it frustrating to work with a "black box" where you can't easily inspect the intermediary stages of the build. Also, I haven't seen significant perf benefits from skipping disk writes; if anything it would make incremental builds less feasible. But this feature is useful, so we should implement it.)

**Comment by Swatinem** [comment]
> ...it would strip `Base` because it found the `@internal` substring in there

puh, that is definitely something that has to be fixed upstream then.

> from an opposite perspective I find it frustrating to work with a "black box" where you can't easily inspect the intermediary stages of the build. Also, I haven't seen significant perf benefits from skipping disk writes; if anything it would make incremental builds less feasible. But this feature is useful, so we should implement it.

Those are definitely valid points. Luckily, when you create small focused libraries, perf is not a problem.

Regarding configuration, I personally would like things to be as zero-config as possible.
`rollup-plugin-dts` comes extremely close to that, since this is basically the *only* thing you need:
https://github.com/Swatinem/rollup-plugin-dts/blob/c1ee0801753e78bb76b2eae0903836b993a983a7/rollup.config.js#L32-L39

Also related to that, the intermediate artifacts themselves are not the problem. Its rather that you have to manage those yourself.

To my own shame, I also didn’t yet understand how to correctly work with TS project references. At work we have a codebase of ~2700 ts files (a lot of those come from external dependencies) and quickly growing. Typechecking times (and IDE performance) are already a pain that I would like to solve somehow.

**Comment by aleclarson** [comment]
I just found [dts-generator](https://github.com/SitePen/dts-generator) by SitePen.

cc @dylans Could you explain how your package differs from the others mentioned in this thread? 😄

**Comment by Swatinem** [comment]
BTW, I also did a really quick research around this topic and I found https://github.com/TypeStrong/dts-bundle, but the latest commit was 2 years ago so this seems pretty dead to me.
I also saw a webpack plugin for types, but can’t remember what it was.

**Comment by timocov** [comment]
There is also https://github.com/nomaed/dts-builder.

**Comment by vytenisu** [comment]
Hello there!

Let me also hop in with [npm-dts](https://www.npmjs.com/package/npm-dts) module I have put together to make it easy with some use cases.

Basically I was a user of dts-generator and encountered a problem at some point (https://github.com/SitePen/dts-generator/issues/121). After trying to solve it for a while I decided to write a simple tool myself which would have less mandatory command-line arguments and would work for my use case.

I tried to make this tool as simple as possible to make it easy to maintain. Added logging with configurable log levels to make it easier to understand what is happening. And well, tried to keep it's source tidy as much as it was viable.

My use case was to pre-bundle several npm libraries (excluding sub-dependencies of course) and generate index.d.ts file for each of those library bundles so that I would still have full TS support when consuming them even though initial src was not uploaded during publishing.

**Quick summary based on discussion points above:**

- Configuration - kept to absolutely minimum/simple set of options. For my case instead of "dts-generator --name ter --project ./ --out index.d.ts --main ter/index" i could simply write "npm-dts generate". It generates index.d.ts (very similar to dts-generator) which is automatically picked up by TS when importing.

- Project compilation during generation - project is compiled to system tmp directory (selected using tmp package based on OS) using TSC. Temporary files are cleaned once done. Basically tool generates typings using TSC itself and then it concatenates + adds some additional stuff resulting in index.d.ts file.

- Following dependencies - npm-dts does not follow/inline dependencies (on purpose) - it simply uses what is generated by TSC. Not sure why I would need to follow dependencies (at least for my use case). Every library exports it's own index.d.ts so if type is found in sub-dependency, that library has it's own index.d.ts and everything works just fine. Library should not bundle-in sub-dependencies because consumers of such library risk of having duplicated code and large bundles as a result. Therefor bundling typings together also does not seem to be necessary.

- Adding custom tags in code (like @public, etc.) - not needed and not supported. Code is parsed using TSC. Whatever TypeScript understands, npm-dts should also.

- Creation of additional files/folders in target project - npm-dts only generates index.d.ts. All temporary files are stored in tmp of the system and is deleted at once after operation is complete. In case it would miserably fail for whatever reason, files would be deleted next time.

- Should it be dep or dev-dep - it should be dev-dependency in most cases unless you would be extending the script with some additional features of your own by wrapping it.

- What TypeScript version is used? - npm-dts uses same TS version as target project. I have not tested which versions of TS are working tbh but as long as this command line is supported "tsc --declaration --emitDeclarationOnly --declarationDir .........", it should be ok.

- Do I need to specify which files to bundle? - no. npm-dts extracts typings from all d.ts files and concatenates them into a single index.d.ts.

- Number of entry points - npm-dts is for single-entry-point libraries. Entry point is determined by "main" key under package.json file of target project.

It also has a webpack plugin and can be executed from another JS script as a dependency. If it works for you, feel free to use it. If not - use dts-generator :) It is a great tool. Aaaand feel free to leave feedback, especially if something does not work as expected.

**Comment by gavar** [comment]
> Hi, and let me also chime in as the maintainer of [rollup-plugin-dts](https://github.com/Swatinem/rollup-plugin-dts)
> 
> As I have also mentioned in my README, I have tried `api-extractor` as well, but my biggest hurdle was also the configuration as @timocov also mentioned. In the end I couldn’t get it to work at all and just gave up.
> 
> So I have nothing to actually compare the _output quality and differences_, like @timocov asked me to here: [Swatinem/rollup-plugin-dts#5](https://github.com/Swatinem/rollup-plugin-dts/issues/5)
> 
> But apart from the configuration issue that was already mentioned, my biggest beef is that it requires me to create intermediate output in the form of pre-rollup `.d.ts` files (point **2.** of @timocov’s list), which goes against how I would like my library workflows to work. (use `tsc` for typechecking only, and use a bundler like rollup / webpack to actually generate the output; I don’t want to have to manage/clean any intermediate build artifacts)
> 
> Also concerning **4.**, I recently found out that `tsc` has a `compilerOptions` for that which I plan to rely on at some point ([Swatinem/rollup-plugin-dts#2](https://github.com/Swatinem/rollup-plugin-dts/issues/2))

`rollup-plugin-dts` is the best option I found, IMHO.
The only missing part I spent a lot of time on was how to properly use it with `rollup-plugin-typescript2`
I would suggest minimal config below - `rollup` execute arrays of configs sequentially (which I haven't found documented well), so it first builds code to common directory and them create a `.d.ts` bundle from it.
```js
import dts from "rollup-plugin-dts";
import rts2 from "rollup-plugin-typescript2";

export default [
  { // TypeScript
    input: "src/index.ts",
    output: [
      {file: "lib/index.js", format: "cjs"},
      {file: "lib/index.mjs", format: "esm"},
    ],
    plugins: [
      rts2({
        cacheRoot: "out/.rts2_cache",
        useTsconfigDeclarationDir: true,
        tsconfigOverride: {
          compilerOptions: {
            declarationDir: "out/.dts",
          },
        },
      }),
    ],
  },
  { // TypeScript Type Definitions
    input: "out/.dts/index.d.ts",
    plugins: [dts()],
    output: {file: "lib/index.d.ts", format: "esm"},
  },
];
```

**Comment by timocov** [comment]
> rollup-plugin-dts is the best option I found, IMHO.

@gavar can you please tell what features/abilities you didn't find in other tools (like api-extractor or dts-bundle-generator), but they are in `rollup-plugin-dts`? I believe it might be helpful for other people who'll see this thread.

**Comment by Swatinem** [comment]
I also recently found a lot more issues with my tool, which are due to the way typescript works:

* in TS, you can override and “monkeypath” things in the global scope, which is not supported, since my tools only works with things you *explicitly* `import`/`export`.
* in TS, you have declaration merging, the most obvious one is when you define overrides for one function, but you can also merge functions with `namespace` declarations, which is the typescript way to attach static properties to functions. When you have multiple declarations with the same name, things can blow up in surprising ways -_-

I still hope that this usecase will be implemented in typescript properly, which would make *all* of these tools obsolete.

**Comment by octogonz** [comment]
API Extractor has fairly comprehensive support for declaration merging.

**Comment by gavar** [comment]
> @gavar can you please tell what features/abilities you didn't find in other tools

## rollup-plugin-dts
### pros
+ supports rollup
+ super fast, takes only 60ms in my case
+ does not have problems with declaration merging (as was stated), all functions with the same name were exported as expected, as well as namespaces 
### cons
- does not support `import {A as B}`, renames back automatically
- renames imports to `[name]$1` when same name occurs twice (relates to above)
- does not generate `/// <reference types="..." />` (don't know if it should)
- places all exports at the end of the file, so it hard to know what's actually exported at a glance

## dts-bundle-generator
### pros
+ exports as it goes
+ generates `/// <reference types="..." />` (don't know if it's useful)
### cons
- does not support `import {A as B}` as defined in docs, I still believe it's possible
- very slow, takes more time than building code itself (compiles code 3 times as I found)
- to few public API so it hard to use with rollup (I've made a [plugin](https://github.com/gavar/wrench/tree/master/packages/roll/src/plugins/dts-bundle-generator.ts) though)
- does not allow to specify compiler options, only tsconfig.json path
- does not allow to turn off compilation of already compiled files (don't know why it's designed like that)
- need configuration for every entry point separately (why not provide allowed libraries for all?) 

## API Extractor
### pros:
+ nearly perfect output
+ very good output formatting (I've made a [dts-pretty](https://github.com/gavar/wrench/tree/master/packages/roll/src/plugins/dts-pretty.ts) for same output, so it does not really affects on my feedback, but it's nice to have out of the box)
+ does pretty much a lot of the stuff most of us not going to use
+ performance is okay, don't know how it would go with watch mode
+ checks JSDocs if you're doing something wrong (basically donne by linters)
### cons
- generated two imports from the same module
- configuration is a pain
- does not allow to configure behavior, only to turn features on/off (this is huge cons)

**Comment by octogonz** [comment]
As of version 7 (which introduced `api-extractor init` for setting up the config file), I'm not aware of people having trouble with the configuration. If you have specific suggestions for improvement, we'd be interested to hear your feedback and see if it can be addressed. If that's off topic for this thread, feel free to open a separate GitHub issue for the project.

**Comment by octogonz** [comment]
>     * generated two imports from the same module

If you have a repo for this,  please open an issue reporting the bug. As far as I know it should not happen unless the two symbols are not provably identical (e.g. imported `React` versus the global ambient declaration with the same name).

**Comment by Swatinem** [comment]
>     * does not generate `/// <reference types="..." />` (don't know if it should)

It should preserve these if the typescript compiler itself generates them.
There is a testcase for it here: https://github.com/Swatinem/rollup-plugin-dts/blob/master/tests/testcases/preserve-reference-directives/expected.d.ts
But possible it’s broken for some cases. Can you reproduce that?

All the other cons you listed is just the way rollup works, its also the same for JS files.

And for the declaration merging, it kind of works, but only if the declarations to be merged are right next to each other.

**Comment by gavar** [comment]
> It should preserve these if the typescript compiler itself generates them.

I meant that `dts-bundle-generator` add those directives as it detects type usages from `@types` directory.

Although @Swatinem, I did found an issue in `rollup-plugin-dts`, when I try to `typeof import`.
Please refer to [issue #49](https://github.com/Swatinem/rollup-plugin-dts/issues/49)

**Comment by gavar** [comment]
worth to mention [wessberg/rollup-plugin-ts](https://github.com/wessberg/rollup-plugin-ts) did a great job.

**Comment by jeremyben** [comment]
Here's my take on it : [bundle option with tsc-prog](https://github.com/jeremyben/tsc-prog#bundle-type-definitions-%EF%B8%8F).

Faster than others because it bundles in memory, by intercepting emitted ".d.ts" files during the initial build. `tsc-prog` is suited for your production builds, as it does not support the `watch` option.

Just to name a few of the corner cases I tried to handle properly: name conflict (with global symbols and local ones), namespace import and property re-export, export from external libraries and json files, the `import` type (with `typeof` as well), etc.

**Comment by timocov** [comment]
> does not support import {A as B} as defined in docs, I still believe it's possible

Yeah, since v4.0.0 it's possible now :)

**Comment by adventure-yunfei** [comment]
User for api-extractor here.
I have to say, if consider dts rollup usage only, `api-extractor` may not support much edge cases as other tools listed here, with the saming effort. It is limited **by its design**, re-creating nearly an entire intermediate ast tree. `api-extractor` has to handle each syntax case that it wants to support, while other dts rollup tools may only handle cases that they don't want to support. And it's true that I met so many unsupported cases that I had to fork the repo to fix.
But `api-extractor` is still a very nice tool. The huge advantage which seems not to be mentioned here, is that **`api-extractor` supports different kind of outputs**, and dts rollup is only one of them. The analyzer in `api-extractor` extract intermediate ast tree from the source to hide details, that's why multiple generators are allowed. It is supported **by its design**.
For now, `api-extractor` tool system supports (see [api-extractor site](https://api-extractor.com/) for details):
- dts rollup
- api document generation
- api review

My recomendation:
- if you only need dts rollup feature, and wish it support as many cases as possible, pick a dts rollup tool (with fast response for issues, or well-designed codes. Trust me, you will still get some edge cases that they don't support.).
- if you're also suprised with other features around d.ts, pick `api-extracor`

**Comment by Shigma** [comment]
I would really like to use dts-bundle-generator but I used a monorepo, which means the bundler I choose should support composite project. It seems that dts-bunder-generator doesn't. I wonder which library above does.

**Comment by loreanvictor** [comment]
Good thread. However, I am searching for the proper tool for a situation where I need to bundle all types from external dependencies as well, with proper name-conflict resolution. Since these are areas where all the mentioned tools seem to be outright avoiding or not guaranteeing proper functionality, can I ask for any recommendations / suggestions?

**Comment by timocov** [comment]
> where I need to bundle all types from external dependencies as well

When it would be needed? Can you please elaborate why do you need this (with examples if possible)?

**Comment by timocov** [comment]
> where I need to bundle all types from external dependencies as well

@loreanvictor when and where it would be needed? Can you please elaborate why do you need this (with examples if possible)?

**Comment by loreanvictor** [comment]
@timocov I need it for [this project](https://github.com/loreanvictor/tyfon). Basically, I'm wrapping some TypeScript functions automatically into API endpoints, and then transferring the type definitions of these functions to the client-side for auto-generating an SDK for said API. The type definitions will sit alongside some auto-generated `.js` code that basically calls the API endpoint corresponding each of aforementioned functions, making the functions _look_ exactly as they were defined on the API code (while the implementation is different).

This works pretty smoothly except for situations where the _API functions_ rely on types from dependencies, as in this situation the client would also need to have the dependencies installed (or at least, the corresponding types installed). Current work-around is avoiding such types in the _API functions_, which is not ideal as it bars re-using recurring types. The alternative would be to also include all types that the _API functions_ are dependent on in the bundle automatically.

P.s. on revisiting my comment I see that it can be read as "collecting all types exported from all dependencies in a blanket manner and add them to the bundle", but thats not what I meant. I basically need to create type-bundles with zero external dependencies.

**Comment by Swatinem** [comment]
So I think [rollup-plugin-dts](https://github.com/Swatinem/rollup-plugin-dts) should be up for the task. Although I absolutely do not recommend it, I know that a few people use it to bundle up all external dependencies as well, and it will happily take care of name conflicts and everything.

**Comment by adventure-yunfei** [comment]
> Good thread. However, I am searching for the proper tool for a situation where I need to bundle all types from external dependencies as well, with proper name-conflict resolution. Since these are areas where all the mentioned tools seem to be outright avoiding or not guaranteeing proper functionality, can I ask for any recommendations / suggestions?

`api-extractor` supports that with [`bundledPackages`](https://api-extractor.com/pages/commands/config_file/#bundledpackages) option.

**Comment by loreanvictor** [comment]
> Although I absolutely do not recommend it

@Swatinem may I ask why?

**Comment by loreanvictor** [comment]
> > Good thread. However, I am searching for the proper tool for a situation where I need to bundle all types from external dependencies as well, with proper name-conflict resolution. Since these are areas where all the mentioned tools seem to be outright avoiding or not guaranteeing proper functionality, can I ask for any recommendations / suggestions?
> 
> `api-extractor` supports that with [`bundledPackages`](https://api-extractor.com/pages/commands/config_file/#bundledpackages) option.

@adventure-yunfei yeah it seems that similar to dts-bundle-generator, you can only specify a few select external libraries and it will bundle types exported from them (i.e. I could not find the `"all"` option).

**Comment by Swatinem** [comment]
> @Swatinem may I ask why?

Its just that some `@types` packages are manually written and have some really strange idioms in them. What I don’t support, and quite a lot of `@types` use for example is [`export = …`](https://github.com/Swatinem/rollup-plugin-dts/issues/93) syntax. Also bundling foreign code is basically against the whole idea of having packages and sharing/deduping dependencies.

**Comment by loreanvictor** [comment]
> > @Swatinem may I ask why?
> 
> Its just that some `@types` packages are manually written and have some really strange idioms in them. What I don’t support, and quite a lot of `@types` use for example is [`export = …`](https://github.com/Swatinem/rollup-plugin-dts/issues/93) syntax. Also bundling foreign code is basically against the whole idea of having packages and sharing/deduping dependencies.

makes sense. tbh, if there was a way to just install types of a package, the cleanest solution for my case would also be simply installing all types as explicit dependencies of the client SDK instead of bundling them all together.

question though: since this pitfall affects mostly `@types` packages, and since `@types` package can actually be safely installed in the client environment as explicit dependencies, is there a way to exclude paths / libraries / etc from `rollup-plugin-dts`'s `"respectExternal"`?

**Comment by Swatinem** [comment]
when using `respectExternal`, you can use the normal `external` setting from rollup.
